lib/phoenix/channel/driver.ex

```diff
@@ -0,0 +1,252 @@
+# TODO(sj): move the file to the proper place
+defmodule Phoenix.Socket.Driver do
+  @moduledoc """
+  Implementation of the Phoenix Channels protocol.
+
+  The Phoenix Channels protocol describes how to hold multiple separate conversations
+  (channels) over a single persistent connection (socket). The connection is
+  powered by transports, such as websocket or long polling, while this module
+  provides the stock implementation of the protocol, with following properties:
+
+    * Each channel runs in a separate process.
+
+    * Channel processes are direct children of the socket process.
+
+    * Termination of a channel process doesn't affect other channel processes or
+      the socket process.
+
+    * If the socket process terminates, all channel processes will be stopped as well,
+      regardless of the exit reason.
+
+  This module was not meant to be used directly. Instead you can pass it as an option
+  to the transport (TODO(sj): explain how, once we settle on the approach). This
+  happens by default when your module is powered by `Phoenix.Socket`.
+  """
+  @behaviour Phoenix.Transports.Driver
+
+  require Logger
+  alias Phoenix.Socket
+  alias Phoenix.Socket.Broadcast
+  alias Phoenix.Socket.Message
+  alias Phoenix.Socket.Reply
+
+  @client_vsn_requirements "~> 1.0"
+  @protocol_version "1.0.0"
+
+  @doc """
+  Returns the Channel Transport protocol version.
+  """
+  def protocol_version, do: @protocol_version
+
+  @doc false
+  def transports(endpoint, driver_opts) do
+    alias Phoenix.Transports
+
+    socket_handler = Keyword.fetch!(driver_opts, :socket_handler)
+    path = Keyword.fetch!(driver_opts, :path)
+
+    for {transport_name, {transport, transport_opts}} <- socket_handler.__transports__ do
+      {
+        Path.join(path, Atom.to_string(transport_name)),
+        %Transports.Driver.Config{
```

Comment:
Both `socket_module` and `transport_name` *are* specific to this driver implementation. Since your work is exactly in abstracting them, I don't think they should be required. Here is my proposal:

    @spec driver.transports(endpoint, path, opts) ::
      {path, transport_module, transport_options}

The only reason to return `driver_opts` is if the transport will now explicitly pass a subset of its options to driver, if we are going to support that, then I am OK with returning `{path, transport_module, transport_opts, driver_opts}`. If we support `driver_opts`, we don't need to pass the endpoint to the driver because it can be added to `driver_opts`.